File: apps/api/src/domain/template/transaction_template.ts
import { AggregateRoot, Criteria, Paginate } from "@abejarano/ts-mongodb-criteria";

export type TransactionTemplatePrimitives = {
  id?: string;
  templateId: string;
  userId: string;
  name: string; // "Netflix", "Aluguel", etc.
  type: "income" | "expense" | "transfer";
  amount: number;
  currency: string;
  categoryId?: string;
  fromAccountId?: string;
  toAccountId?: string;
  note?: string;
  tags?: string[];
  createdAt: Date;
  updatedAt: Date;
};

export class TransactionTemplate extends AggregateRoot {
  constructor(private readonly props: TransactionTemplatePrimitives) {
    super();
  }

  static fromPrimitives(props: TransactionTemplatePrimitives): TransactionTemplate {
    return new TransactionTemplate(props);
  }

  toPrimitives(): TransactionTemplatePrimitives {
    return { ...this.props };
  }

  get id(): string | undefined {
    return this.props.id;
  }

  get templateId(): string {
    return this.props.templateId;
  }

  getId(): string {
    return this.props.id ?? this.props.templateId;
  }

  get userId(): string {
    return this.props.userId;
  }

  get name(): string {
    return this.props.name;
  }
}

export interface TransactionTemplateRepository {
  upsert(template: TransactionTemplate): Promise<void>;
  list(criteria: Criteria): Promise<Paginate<TransactionTemplate>>;
  byId(templateId: string): Promise<TransactionTemplate | undefined>;
  remove(templateId: string): Promise<void>;
}

File: apps/flutter/lib/features/templates/domain/entities/transaction_template.dart
class TransactionTemplate {
  final String id;
  final String userId;
  final String name;
  final double amount;
  final String type; // 'income', 'expense', 'transfer'
  final String currency;
  final String? categoryId;
  final String? fromAccountId;
  final String? toAccountId;
  final String? note;
  final List<String>? tags;

  const TransactionTemplate({
    required this.id,
    required this.userId,
    required this.name,
    required this.amount,
    required this.type,
    required this.currency,
    this.categoryId,
    this.fromAccountId,
    this.toAccountId,
    this.note,
    this.tags,
  });

  factory TransactionTemplate.fromJson(Map<String, dynamic> json) {
    return TransactionTemplate(
      id: json['templateId'] ?? json['id'],
      userId: json['userId'],
      name: json['name'],
      amount: (json['amount'] as num).toDouble(),
      type: json['type'],
      currency: json['currency'],
      categoryId: json['categoryId'],
      fromAccountId: json['fromAccountId'],
      toAccountId: json['toAccountId'],
      note: json['note'],
      tags: (json['tags'] as List?)?.map((e) => e as String).toList(),
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'id': id,
      'userId': userId,
      'name': name,
      'amount': amount,
      'type': type,
      'currency': currency,
      'categoryId': categoryId,
      'fromAccountId': fromAccountId,
      'toAccountId': toAccountId,
      'note': note,
      'tags': tags,
    };
  }
}
